raw-ipa-lib/src/user.rs

```diff
@@ -253,4 +276,60 @@ mod tests {
             User::point_from_matchkey(OTHER_PROVIDER, MATCHKEY.as_bytes())
         );
     }
+
+    #[test]
+    fn test_generate_event_report() {
+        const PROVIDER_1: &str = "social.example";
+        const PROVIDER_2: &str = "email.example";
+        const PROVIDER_3: &str = "news.example";
+        const MATCHING_MATCHKEY: &str = "this_one_matches";
+
+        let mut rng = thread_rng();
+        let d1 = ThresholdDecryptionKey::new(&mut rng);
+        let d2 = ThresholdDecryptionKey::new(&mut rng);
+        let tek = ThresholdEncryptionKey::new(&[d1.encryption_key(), d2.encryption_key()]);
+
+        let providers = [PROVIDER_1, PROVIDER_2, PROVIDER_3];
+
+        let mut u1 = User::new(0, tek);
+        u1.set_matchkey(PROVIDER_1, MATCHING_MATCHKEY);
+        u1.set_matchkey(PROVIDER_2, "something_random");
+        u1.set_matchkey(PROVIDER_3, "also_very_random");
+
+        let mut u2 = User::new(1, tek);
+        u2.set_matchkey(PROVIDER_1, MATCHING_MATCHKEY);
+        u2.set_matchkey(PROVIDER_2, "does_not_match");
+        u2.set_matchkey(PROVIDER_3, "also_does_not_match");
+
+        let r1 = u1.generate_event_report(&providers);
+        let r2 = u2.generate_event_report(&providers);
+
+        // None combination of encrypted match keys should match
+        assert_ne!(r1.encrypted_match_keys.get(PROVIDER_1), r2.encrypted_match_keys.get(PROVIDER_1));
+        assert_ne!(r1.encrypted_match_keys.get(PROVIDER_1), r2.encrypted_match_keys.get(PROVIDER_2));
+        assert_ne!(r1.encrypted_match_keys.get(PROVIDER_1), r2.encrypted_match_keys.get(PROVIDER_3));
+
+        assert_ne!(r1.encrypted_match_keys.get(PROVIDER_2), r2.encrypted_match_keys.get(PROVIDER_2));
+        assert_ne!(r1.encrypted_match_keys.get(PROVIDER_2), r2.encrypted_match_keys.get(PROVIDER_3));
+
+        assert_ne!(r1.encrypted_match_keys.get(PROVIDER_3), r2.encrypted_match_keys.get(PROVIDER_3));
+
+        let fully_decrypted_r1: HashMap<_, _> = r1.encrypted_match_keys.iter().map(
+            |(p, emk)| (p.to_string(), d2.decrypt(d1.threshold_decrypt(*emk)))
+        ).collect();
+
+        let fully_decrypted_r2: HashMap<_, _> = r2.encrypted_match_keys.iter().map(
+            |(p, emk)| (p.to_string(), d2.decrypt(d1.threshold_decrypt(*emk)))
+        ).collect();
```

Comment:
It seems like you want a real object to hold this result.  That would allow for the logic to be held somewhere (other than tests) as well.  That object can have a PartialEq implementation that returns true if any matchkey matches (no Eq implementation though, as that doesn't make sense for that type).  Then all you would need to do is take the encrypted matchkeys, decrypt into this new object, and compare.  (More extensive checking is probably useful to be sure that everything is working, of course; see other comments.)

Reply:
That's pretty neat! I like this idea.